src/main/scala/dependencies/ModelFactory.scala

```diff
@@ -52,24 +52,24 @@ object ModelFactory {
       * Take this opportunity to also squash dependencies for the same artifact at different versions
       * down to a single version (the highest).
       */
-    val thirdParty: Set[MavenDependency] = {
-      def flattenDependencies(root: MavenDependency): Seq[MavenDependency] =
+    val thirdParty: Set[MavenDependency[MavenArtifactCoordinates]] = {
+      def flattenDependencies(root: MavenDependency[ArtifactCoordinates]): Seq[MavenDependency[MavenArtifactCoordinates]] =
         if (root.coordinates.isInstanceOf[MavenArtifactCoordinates]) {
-          Seq(root) ++ root.dependsOn.toSeq.flatMap(flattenDependencies)
+          Seq(root.asInstanceOf[MavenDependency[MavenArtifactCoordinates]]) ++ root.dependsOn.toSeq.flatMap(flattenDependencies)
         } else {
           root.dependsOn.toSeq.flatMap(flattenDependencies)
         }
-
+      
       roots.flatMap(t => flattenDependencies(t._2))
-          .groupBy(d => d.coordinates match {
-            case m: MavenArtifactCoordinates => m.artifact
-            case _ => throw new IllegalStateException("Should not see non-maven coordinates here: " + d.coordinates)
-          })
+          .groupBy(_.coordinates.artifact)
           .values
           .flatMap(versionedDependencies => versionedDependencies.size match {
             case 0 => Seq()
             case 1 => Seq(versionedDependencies.head)
-            case _ => Seq(versionedDependencies.maxBy(_.coordinates.asInstanceOf[MavenArtifactCoordinates].semanticVersion))
+            case _ =>
+              val newest = versionedDependencies.maxBy(_.coordinates.semanticVersion)
+              System.err.println(s"Pinning to latest version ${newest.coordinates.version} of ${newest.coordinates.artifact} from amongst versions: ${versionedDependencies.map(_.coordinates.version)}")
```

Comment:
Gradle only rationalizes dependencies _within_ a given target.  Across targets, versions can be all over the place due to different transitive dependency graphs.  

I'm canonicalizing the _entire_ transitive dependency graph for all projects in a multi-project repo, so this is needed to squash everything down to a single (and, preferably, the most recent) version.